### PR TITLE
DCOS-12673/fix_azure_flaky

### PR DESCRIPTION
**Pull request by orsenthil**
Attempts to fix certain infrastructure flakiness with Azure.

The main change is this.

```
@retry(retry_on_exception=retry_if_assertion_error, stop_max_attempt_number=45)
```

Rest of the changes are only cursory improvements in the code that I touched.

Issues

* [DCOS-12673](https://mesosphere.atlassian.net/browse/DCOS-12673)

Checklist

[ ] Included a test which will fail if code is reverted but test is not
[ ] Followed relevant code rules Rules for Packages and Systemd

### CODE REVIEW

**Comment by orsenthil** [comment]
This is ready for review @lingmann , @branden , and @cmaloney .

File: test_util/azure_test_driver.py [hunk 1]
```diff
@@ -124,14 +120,17 @@ def main():
         print("Creating template deployment ...")
         deploy_poller = rmc.deployments.create_or_update(group_name, deployment_name, deployment_properties)
 
-        # Stop after 45 attempts (each one takes up to one minute)
-        @retry(stop_max_attempt_number=45)
+        def retry_if_assertion_error(exception):
+            """Return True if we should retry (in this case when it's an AssertionError), False otherwise"""
+            return isinstance(exception, AssertionError)
+
+        @retry(retry_on_exception=retry_if_assertion_error, stop_max_attempt_number=45)
```
Comment (BenWhitehead) [inline test_util/azure_test_driver.py]: How frequently will the retry be attempted? I see it's limited to 45 attempts, but not sure on the retry interval.

Comment (orsenthil) [inline test_util/azure_test_driver.py]: The retry interval is defined by the thread wait time and assertion in the below.

```
             res = deploy_poller.result(timeout=60)
             ...
             assert deploy_poller.done(), "Not done deploying."		    
```

So, after every 60 seconds and assertion is made if the task is done and this is done for 45 times.

* Previously, the 45 times retry was done for any failure.
* Now, this retry of task is done only only carried out for `AssertionError`, any other Infrastructure issues like `CloudError` will fail fast.

Hope this helps.

Comment (BenWhitehead) [inline test_util/azure_test_driver.py]: got it, I figured the `timeout=60` meant the timeout for the rpc to azure had to return within 60 seconds.

Thanks for the clarification.

Comment (amitaekbote) [inline test_util/azure_test_driver.py]: Was the flakiness caused by AssertionError and the fact that we didnt retry on that particular failure?

Comment (orsenthil) [inline test_util/azure_test_driver.py]: Actually, retrying on AssertionError only is the right thing to do. The way the code is written, AssertionError is part of the logic for checking if the task is done and a retry is a while loop to for waiting until the task is complete. 

The unfortunate scenario when CloudError on Azure happens and it kills the task, we should not be retrying for 45 times, the clouderror must have killed the long running task and we should bail out then and exit with a failure.  This our fix (fail fast) is the fix the flaky azure tests which are not in our control.

HTH.

Comment (amitaekbote) [inline test_util/azure_test_driver.py]: makes sense, lgtm

**Pull request merged**